### PR TITLE
resolver: autocomplete api

### PR DESCRIPTION
**Pull request by mcutalo88**
### Description

This add the resolver autocomplete api, with the initial implantation is just for AWS. The Kubernetes resolver will come in a follower up pr as they require a more complex solution. I also introduced some basic testing for the `resolver/aws` package and for this feature.

Enabling autocomplete is done automatically by the backend, if the topology service is enabled and all of its dependencies are satisfied, we enabled all resources to be autocompleteable. This does mean that resource which are not yet implemented for autocomplete will still send queries to the API, however support will come soon after this initial work is in. In addition moving forward we can invest in the feature flag API https://github.com/lyft/clutch/issues/4 for a cleaner signal to the frontend.

The original idea was to set metadata on the resolver protos to denote which resource was autocompleteable. However we still needed to modify this value on the backend if the topology service and its dependencies were not satisfied. This configuration was misleading as the user would expect autocomplete to be enabled even tho it could change behind the scenes. But did enabled us to turn on autocomplete per resource rather than holistically. 
```proto
message InstanceID {
  option (clutch.resolver.v1.schema) = {
    display_name : "instance ID"
    search : {
       enabled : true
       is_autocompleteable : true <-- original idea
    }
  };
}
  ```

### Testing Performed
Unit / Local.

```
curl --location --request POST 'http://localhost:8080/v1/resolver/autocomplete' \
--header 'Content-Type: application/json' \
--data-raw '{
    "want": "type.googleapis.com/clutch.aws.ec2.v1.AutoscalingGroup",
    "search": "kitchensink"
}'

{
    "results": [
        {
            "id": "kitchensink-production",
            "label": ""
        },
        {
            "id": "kitchensink-staging",
            "label": ""
        },
        {
            "id": "kitchensink-canary",
            "label": ""
        }
    ]
}
```

### GitHub Issue

This makes progress on issues https://github.com/lyft/clutch/issues/921 and https://github.com/lyft/clutch/issues/30

### CODE REVIEW

**Comment by mcutalo88** [comment]
cc @lyft/clutch-admin

File: backend/module/resolver/resolver.go
Comment (danielhochman) [inline backend/module/resolver/resolver.go:184]: add comment

File: backend/resolver/aws/aws.go
Comment (danielhochman) [inline backend/resolver/aws/aws.go:156]:
```suggestion
		return nil, fmt.Errorf("to use the autocomplete api you must first setup the topology service")
```

File: backend/resolver/aws/aws.go
Comment (danielhochman) [inline backend/resolver/aws/aws.go:178]: nit: make() with preallocation

File: api/resolver/v1/resolver_api.proto
Comment (danielhochman) [inline api/resolver/v1/resolver_api.proto:37]: nit: autocomplete is one word, please change throughout
```suggestion
  rpc Autocomplete(AutocompleteRequest) returns (AutocompleteResponse) {
```

File: backend/resolver/resolver.go
Comment (danielhochman) [inline backend/resolver/resolver.go:22]: i would go ahead and make this an optional parameter that threads through and then default it to this value at the module level rather than using a const

File: backend/resolver/resolver.go
Comment (danielhochman) [inline backend/resolver/resolver.go:45]: []string is too limited. we may want to show additional metadata about a selection in the autocomplete pane.

Screenshot from 2021-02-08 12-22-05

File: api/resolver/v1/resolver_api.proto
Comment (danielhochman) [inline api/resolver/v1/resolver_api.proto:53]: see comment in resolver.go about return type

File: backend/resolver/aws/aws.go
Comment (danielhochman) [inline backend/resolver/aws/aws.go:159]: i think you can probably some of this down into the topology package, but a TODO is good for now. i just don't want us to start copy and pasting without thinking through it when we go to add autocomplete to the next target.

**Review by danielhochman — COMMENTED** [review]
cool stuff, lots of food for thought to talk through mostly centered around the API and future expansion

File: backend/resolver/resolver.go
Comment (mcutalo88) [reply backend/resolver/resolver.go:45]: Yeah true good call, let me the ability to add more information.

File: api/resolver/v1/resolver_api.proto
Comment (mcutalo88) [inline api/resolver/v1/resolver_api.proto:58]: Actually thinking about this more as I implemented the frontend, it might just be better to have a simpler approach like the one below. 

```proto
  message AutocompleteResult {
    string id = 1;
    string label = 2;
  }
```

This will just be easier to map to the text-field autocomplete implementation, as we can render the option like so.
```typescript
renderOption={option => (
  <>
    <span>{option.id}</span>
    {option.label}
  </>
)}
```

File: api/resolver/v1/resolver_api.proto
Comment (mcutalo88) [reply api/resolver/v1/resolver_api.proto]: I just realized this, going to change it back unless there is a way to make the linter happy / ignore this?
```
➜ clutch git:(resolver-autocomplete-api) make api-lint
tools/compile-protos.sh -c "/Users/mcutalo/gocode/src/github.com/lyft/clutch/api" -l
--- /Users/mcutalo/gocode/src/github.com/lyft/clutch/api/resolver/v1/resolver_api.proto
 v1beta1". See https://docs.buf.build/faq for more details.
37:20:RPC request type "AutocompleteRequest" should be named "AutoCompleteRequest" or "ResolverAPIAutoCompleteRequest".
37:50:RPC response type "AutocompleteResponse" should be named "AutoCompleteResponse" or "ResolverAPIAutoCompleteResponse".
make: *** [api-lint] Error 1
```

Comment (danielhochman) [reply api/resolver/v1/resolver_api.proto]: it needs to match everywhere, check :37

Comment (mcutalo88) [reply api/resolver/v1/resolver_api.proto]: ooo i see nice

**Review by mcutalo88 — COMMENTED** [review]
Alright @danielhochman ready for another review, I have update the PR description with some rational.

File: api/resolver/v1/schema.proto
Comment (danielhochman) [inline api/resolver/v1/schema.proto:46]: super nit: plz autocomplete or autocomplete_enabled, autocompletable is prone to spelling errors

File: backend/resolver/resolver.go
Comment (danielhochman) [inline backend/resolver/resolver.go:41]:
```suggestion
	Autocomplete(ctx context.Context, typeURL, search string, limit uint64) ([]*resolverv1.AutocompleteResponse_AutocompleteResult, error)
```

File: api/resolver/v1/resolver_api.proto
Comment (danielhochman) [inline api/resolver/v1/resolver_api.proto:50]: nit: keep consistent with other APIs

```suggestion
  uint64 limit = 3;
```

File: api/resolver/v1/resolver_api.proto
Comment (danielhochman) [inline api/resolver/v1/resolver_api.proto:56]: comments on all fields please

File: backend/module/resolver/resolver.go
Comment (danielhochman) [inline backend/module/resolver/resolver.go:202]: i'm not sure if this is correct. we want to autocomplete from any resolver that can provide the object, right?

File: backend/resolver/aws/aws.go
Comment (danielhochman) [inline backend/resolver/aws/aws.go:74]: nit: combine, harder to read this way toplogyService vs toplogySvc

File: backend/resolver/aws/aws.go
Comment (danielhochman) [inline backend/resolver/aws/aws.go:178]: you probably want your own default limit that's lower than whatever the default limit of the topo API is

**Review by danielhochman — COMMENTED** [review]
mostly nits but a couple of questions to work through

File: backend/module/resolver/resolver.go
Comment (mcutalo88) [reply backend/module/resolver/resolver.go:202]: Yeah good question, for the initial implementation I was going to leave this out until we decided on how we wanted to merge / sort results from different resolvers to surface the most relevant ones to the user. But maybe as a start it's fine if the list is not perfectly sorted and Im overthinking it. 

For now tho I can just merge these results together and will leave a comment as to what additional work needs to be done / thought about.

File: backend/resolver/aws/aws.go
Comment (rithujohn191) [inline backend/resolver/aws/aws.go:77]:
```suggestion
			return nil, errors.New("incorrect topology service type")
```

File: backend/module/resolver/resolver.go
Comment (danielhochman) [reply backend/module/resolver/resolver.go:202]: For now you could just stop if limit satisfied? Later we can do Levenshtein distance or some more advanced algo.

File: backend/resolver/aws/aws.go
Comment (rithujohn191) [inline backend/resolver/aws/aws.go:156]:
```suggestion
		return nil, fmt.Errorf("topology service not initialized. To use the autocomplete api you must first setup the topology service")
```

File: backend/module/resolver/resolver.go
Comment (mcutalo88) [reply backend/module/resolver/resolver.go:202]: Alright yeah good suggestion let me start with that.

File: backend/resolver/aws/aws.go
Comment (mcutalo88) [reply backend/resolver/aws/aws.go:74]: I can rename to make it easier to read, but I needed two variables so I can supply `topologyService` to the resolver struct which can be nil. Or i can just rearrange this logic and move the construction above the topology initialization bits and just assign the value eg: `r.topology = topologyService`.

Comment (danielhochman) [reply backend/resolver/aws/aws.go:74]: yes i meant combine the if to make the scope clearer
```suggestion
	if svc, ok := service.Registry[topology.Name]; ok {
```

Comment (mcutalo88) [reply backend/resolver/aws/aws.go]: Oh ok thanks for the clarification I didn't understand at first

File: backend/resolver/aws/aws.go
Comment (mcutalo88) [reply backend/resolver/aws/aws.go:178]: Yeah true let me do so

File: backend/clutch-config.yaml
Comment (scarlettperry) [inline backend/clutch-config.yaml:37]: were these changes just for local testing?

File: api/resolver/v1/resolver_api.proto
Comment (scarlettperry) [inline api/resolver/v1/resolver_api.proto:54]: super nit: if you think you'll need `AutocompleteResult` elsewhere, may want to move it out `AutocompleteResponse`

File: backend/resolver/aws/aws.go
Comment (scarlettperry) [inline backend/resolver/aws/aws.go:163]: should we check the limit before using the default limit? or maybe you are checking somewhere else and im misunderstanding the flow.

**Review by scarlettperry — APPROVED** [review]
lgtm! jst a Q and a nit

File: backend/clutch-config.yaml
Comment (mcutalo88) [reply backend/clutch-config.yaml:37]: oops yeah ill remove all my changes here

File: api/resolver/v1/resolver_api.proto
Comment (mcutalo88) [reply api/resolver/v1/resolver_api.proto:54]: Yeah i was thinking about this, maybe i'll just move it out just incase? But nothing comes to mind at the moment.

Comment (dschaller) [reply api/resolver/v1/resolver_api.proto:54]: could be useful to move it out if we want to use it as a type on the frontend

File: backend/resolver/resolver.go
Comment (dschaller) [inline backend/resolver/resolver.go:21]: should we have this be config driven?

File: backend/module/resolver/resolver.go
Comment (dschaller) [inline backend/module/resolver/resolver.go:211]: Shouldn't this use the limit?

File: backend/module/resolver/resolver.go
Comment (dschaller) [inline backend/module/resolver/resolver.go:207]:
```suggestion
			// Append results from all resolvers up to the limit.
```

File: backend/resolver/aws/aws.go
Comment (dschaller) [inline backend/resolver/aws/aws.go:78]: should this be debug?

File: backend/module/resolver/resolver.go
Comment (mcutalo88) [reply backend/module/resolver/resolver.go:211]: good catch i literally wrote the code for this above and did not use it.

File: backend/resolver/resolver.go
Comment (mcutalo88) [reply backend/resolver/resolver.go:21]: Yeah sure we can do that so a user can override the default.

Comment (mcutalo88) [reply backend/resolver/resolver.go:21]: Actually I'm going to hold off on this for now, we dont have any configuration currently in place for the resolver and slotting it in does not seem straight forward at the moment. We can tackle this in a separate PR as its going to overload this one. If it was simply i would of just added it now.

**Review by dschaller — APPROVED** [review]

**Pull request merged**
